### PR TITLE
zlib: Update to 1.2.12

### PR DESCRIPTION
**Pull request by eschnett**
Update to version 1.2.12.
Mark older versions as deprecated because they have security bugs.

### CODE REVIEW

**Review by haampie — DISMISSED** [review]

**Review by haampie — DISMISSED** [review]

**Comment by haampie** [comment]
~Why are changes to `mfem` added?~ sorry, I missed that the issue only shows up later when compiling ascent...

**Comment by eschnett** [comment]
[I posted this to the wrong PR originally... Sorry.]

The change to mfem is intentional. Updating zlib breaks mfem, since mfem accidentally uses the system zlib for linking. This fails when the Spack zlib is newer than the system zlib, since the linker then can't find the right (versioned) symbols.

Since mfem is built as part of CI, the CI checks will not pass without this correction to mfem.

On my system, mfem also tries to link against the wrong MPI library; it uses the system MPI instead of the one built by Spack, for the same reason. This correction to mfem corrects this problem as well.

**Comment by haampie** [comment]
I just downloaded the binaries from the failed build and got this:

```
$ ./libtree_x86_64 libmfem.so.4.3 
libmfem.so.4.3 
├── libHYPRE-2.23.0.so [rpath]
│   ├── libmpi.so.40 [rpath]
│   │   ├── libopen-rte.so.40 [rpath]
│   │   │   ├── libopen-pal.so.40 [rpath]
│   │   │   │   ├── libevent_core-2.1.so.7 [rpath]
│   │   │   │   │   └── libpthread.so.0 [ld.so.conf]
│   │   │   │   ├── libevent_pthreads-2.1.so.7 [rpath]
│   │   │   │   │   └── libpthread.so.0 [ld.so.conf]
│   │   │   │   ├── libhwloc.so.15 [rpath]
│   │   │   │   │   ├── libxml2.so.2 [rpath]
│   │   │   │   │   │   ├── /home/software/spack/.../linux-ubuntu18.04-x86_64/gcc-7.5.0/zlib-1.2.12-uc7zedamvzzmit4o3bvflgtxbnbsxhmr/lib/libz.so [rpath]
│   │   │   │   │   │   ├── liblzma.so.5 [rpath]
│   │   │   │   │   │   │   └── libpthread.so.0 [ld.so.conf]
│   │   │   │   │   │   └── libiconv.so.2 [rpath]
│   │   │   │   │   └── libpciaccess.so.0 [rpath]
│   │   │   │   ├── /home/software/spack/.../linux-ubuntu18.04-x86_64/gcc-7.5.0/zlib-1.2.12-uc7zedamvzzmit4o3bvflgtxbnbsxhmr/lib/libz.so [rpath]
│   │   │   │   ├── libpthread.so.0 [ld.so.conf]
│   │   │   │   ├── libutil.so.1 [ld.so.conf]
│   │   │   │   └── librt.so.1 [ld.so.conf]
│   │   │   │       └── libpthread.so.0 [ld.so.conf]
│   │   │   ├── libevent_core-2.1.so.7 [rpath]
│   │   │   ├── libhwloc.so.15 [rpath]
│   │   │   ├── /home/software/spack/.../linux-ubuntu18.04-x86_64/gcc-7.5.0/zlib-1.2.12-uc7zedamvzzmit4o3bvflgtxbnbsxhmr/lib/libz.so [rpath]
│   │   │   └── libpthread.so.0 [ld.so.conf]
│   │   ├── libopen-pal.so.40 [rpath]
│   │   ├── libhwloc.so.15 [rpath]
│   │   ├── librt.so.1 [ld.so.conf]
│   │   └── libpthread.so.0 [ld.so.conf]
│   └── libopenblas.so.0 [rpath]
│       └── libgfortran.so.4 [rpath of 1]
│           └── libquadmath.so.0 [rpath of 1]
├── libmetis.so [rpath]
├── libconduit.so [rpath]
│   └── libgfortran.so.4 [rpath of 1]
├── libconduit_relay.so [rpath]
│   ├── libconduit_blueprint.so [rpath]
│   │   ├── libconduit.so [rpath]
│   │   └── libgfortran.so.4 [rpath of 1]
│   ├── libconduit.so [rpath]
│   ├── libhdf5.so.200 [rpath]
│   │   ├── libmpi.so.40 [rpath]
│   │   ├── /home/software/spack/.../gcc-7.5.0/zlib-1.2.12-uc7zedamvzzmit4o3bvflgtxbnbsxhmr/lib/libz.so [rpath]
│   │   └── libpthread.so.0 [ld.so.conf]
│   ├── librt.so.1 [ld.so.conf]
│   └── libpthread.so.0 [ld.so.conf]
├── libconduit_blueprint.so [rpath]
├── /home/software/spack/.../linux-ubuntu18.04-x86_64/gcc-7.5.0/zlib-1.2.12-uc7zedamvzzmit4o3bvflgtxbnbsxhmr/lib/libz.so.1 [rpath]
├── libmpi.so.40 [rpath]
├── libpthread.so.0 [ld.so.conf]
└── librt.so.1 [ld.so.conf]
```

seems like it could be picked up in rpaths (as in, the symlink from soname -> file exists), but the library itself doesn't have a soname.

And yeah, the fact that `libz.so.1` is needed means it links to he system version :(.

Other libraries like libxml2 have DT_NEEDED: libz.so.

**Comment by haampie** [comment]
@eschnett it seems when I run configure outside of spack I get a soname...

**Comment by eschnett** [comment]
@haampie If you configure outside Spack, then you are likely using the system compilers and not Spack's compiler wrappers. This might lead to different behaviour.

**Comment by haampie** [comment]
Their configure file is bogus. It does not detect gcc when CC is set. Maybe this weekend I'll turn zlib into a MakefilePackage with our own Makefile, given that `cc *.c -shared` just works...

**Comment by eschnett** [comment]
@haampie The correction to `mfem` is still necessary, independent of how `zlib` is built.

**Review by haampie — APPROVED** [review]

File: var/spack/repos/builtin/packages/zlib/package.py [hunk 1]
```diff
@@ -19,11 +19,10 @@ class Zlib(Package):
     # URL must remain http:// so Spack can bootstrap curl
     url = "https://zlib.net/fossils/zlib-1.2.11.tar.gz"
```
Comment (ax3l) [inline var/spack/repos/builtin/packages/zlib/package.py:20]: I just saw this comment and this URL is not `http://` anymore

Comment (eschnett) [reply var/spack/repos/builtin/packages/zlib/package.py:20]: That's probably best discussed in a new issue.

**Pull request merged**